src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java

```diff
@@ -291,4 +226,91 @@ private String generateWriteErrors(
     }
     return "[" + String.join(", ", errorString) + "]";
   }
+
+  static final class StartedMongoSinkTask {
```

Comment:
The refactoring in `MongoSinkTask` did not change any logic, but merely extracted the instance logic that is not related to the Kafka Connect-specific lifecycle of the `MongoSinkTask` to a new class `StartedMongoSinkTask`, which can now be easily unit tested.

Reply:
I did not make it a top-level class initially, because I was trying to minimize the changes in `MongoSinkTask`. But I am happy to have such a change. Done.

> So this is just a mechanism for setting the mongo client?

Yes it's just manual dependency injection (DI):

- It results in dependencies of `StartedMongoSinkTask` (sink config, MongoDB client, error reporter) being expressed clearly, which improves readability;
- It allows the code that creates such tasks to control the implementation of dependencies, which decouples `StartedMongoSinkTask` from those implementations. This is crucial for unit tests, because they now can easily specify mocked implementations.

> the stages are all still linked to the Kafka lifestyle - eg: stop and put

The important part is that the two aforementioned points are still true. I poorly expressed the idea in the initial comment, but hopefully now it is clearer.

> Is there a cleaner way to achieve that for testing.

I don't think there is. DI is applied here right to its point. And of all different styles of DI, constructor injection is the cleanest one.